walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,130 @@
 
 package com.google.sps;
 
+import com.google.sps.TimeRange;
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    // If request duration is longer than a day
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+
+    // If no one has other events
+    if (events.size() == 0) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    List<TimeRange> mandatoryBusyTimes = getBusyTimes(events, request.getAttendees());
+    List<TimeRange> mandatoryOpenTimes = getOpenTimes(mandatoryBusyTimes, request.getDuration());
+    
+    List<TimeRange> optionalBusyTimes = getBusyTimes(events, request.getOptionalAttendees());
+    List<TimeRange> optionalOpenTimes = getOpenTimes(optionalBusyTimes, request.getDuration());
+
+    //return getOverlappedTimes(mandatoryOpenTimes, optionalOpenTimes, request.getDuration());
+    return mandatoryOpenTimes;
+  }
+
+  // Returns a list of busy times for "requestedAttendees", given other "events"
+  private List<TimeRange> getBusyTimes(Collection<Event> events, Collection<String> requestedAttendees) {
+
+    // Find events including requested attendees and add those time ranges to "busyTimes"
+    List<TimeRange> busyTimes = new ArrayList<TimeRange>();
+    for (Event event: events) {
+      Set<String> attendees = event.getAttendees();
+
+      // Check if "event" includes a requested attendee
+      for (String requestedAttendee : requestedAttendees) {
+        if (attendees.contains(requestedAttendee)) {
+          busyTimes.add(event.getWhen());
+          break;
+        }
+      }
+    }
+    if (busyTimes.size() == 0) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+    Collections.sort(busyTimes, TimeRange.ORDER_BY_START);
+
+    // Create new list of times with no overlapping times
+    List<TimeRange> busyTimesFinal = new ArrayList<TimeRange>();
+    TimeRange thisTime = busyTimes.get(0);
+    TimeRange nextTime;
+    for (int i = 1; i < busyTimes.size(); i++) {
+      nextTime = busyTimes.get(i);
+
+      // If the times overlap, create a new TimeRange that merges them and set "thisTime" to it
+      if (thisTime.overlaps(nextTime)) {
```

Comment:
I feel it would be more extensible if we could create a set of utils:
1. findOverlap() to find intersection of times
2. Extract logic here to find union of times
3. Change getOpenTimes() to find complement of times (so that it makes sense when we have a list of open times, we call it for a list of busy times)

Reply:
What part are you referring to for finding the union of times?

I moved the code for merging the overlapped busy times into a new helper function, "mergeOverlappedTimes()"

And I'm not sure if it makes sense to change getOpenTimes() to find the complement of times because getOpenTimes() doesn't find the exact complement. It only adds the open times if they can fit the meeting duration ("duration" parameter). I'm not sure what this parameter would do if we have a list of open times and call the function for a list of busy times. What do you think?

Reply:
It's just line 107 and 108 I was referring to for finding the union of times. Don't worry about this since we are not finding the intersection now and it's not necessary to create the utils.
For getOpenTimes(), I only meant to change the name, so that it makes sense when we have a list of open times, we call it with duration=0 for a list of busy times. Let's leave it here and only do it if we really need to reuse this function.